File: sdk/go/epoch/webhook.go
package epoch

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// WebhookVerification is the authenticated replay identity of one Epoch
// webhook request. Persist DeliveryID and Attempt before applying side effects.
type WebhookVerification struct {
	DeliveryID string
	Attempt    uint32
	SignedAt   time.Time
}

// VerifyWebhookSignature verifies the exact request body, delivery identity,
// attempt, and timestamp produced by Epoch's signed webhook executor.
func VerifyWebhookSignature(
	secret, body []byte,
	deliveryID, attemptHeader, timestampHeader, signatureHeader string,
	now time.Time,
	tolerance time.Duration,
) (WebhookVerification, error) {
	if len(secret) == 0 {
		return WebhookVerification{}, fmt.Errorf("epoch: webhook secret is required")
	}
	if strings.TrimSpace(deliveryID) == "" {
		return WebhookVerification{}, fmt.Errorf("epoch: webhook delivery ID is required")
	}
	attemptValue, err := strconv.ParseUint(attemptHeader, 10, 32)
	if err != nil || attemptValue == 0 {
		return WebhookVerification{}, fmt.Errorf("epoch: webhook attempt must be a positive uint32")
	}
	if strconv.FormatUint(attemptValue, 10) != attemptHeader {
		return WebhookVerification{}, fmt.Errorf("epoch: webhook attempt must use canonical decimal encoding")
	}
	timestamp, err := strconv.ParseInt(timestampHeader, 10, 64)
	if err != nil || timestamp < 0 {
		return WebhookVerification{}, fmt.Errorf("epoch: webhook timestamp must be a non-negative integer")
	}
	if strconv.FormatInt(timestamp, 10) != timestampHeader {
		return WebhookVerification{}, fmt.Errorf("epoch: webhook timestamp must use canonical decimal encoding")
	}
	if tolerance <= 0 {
		return WebhookVerification{}, fmt.Errorf("epoch: webhook timestamp tolerance must be positive")
	}
	signedAt := time.Unix(timestamp, 0).UTC()
	if delta := now.UTC().Sub(signedAt); delta < -tolerance || delta > tolerance {
		return WebhookVerification{}, fmt.Errorf("epoch: webhook timestamp is outside the allowed tolerance")
	}
	if len(signatureHeader) != len("v1=")+sha256.Size*2 || !strings.HasPrefix(signatureHeader, "v1=") {
		return WebhookVerification{}, fmt.Errorf("epoch: webhook signature must use v1 lowercase hexadecimal")
	}
	encodedSignature := strings.TrimPrefix(signatureHeader, "v1=")
	if encodedSignature != strings.ToLower(encodedSignature) {
		return WebhookVerification{}, fmt.Errorf("epoch: webhook signature must use v1 lowercase hexadecimal")
	}
	provided, err := hex.DecodeString(encodedSignature)
	if err != nil {
		return WebhookVerification{}, fmt.Errorf("epoch: webhook signature must use v1 lowercase hexadecimal")
	}
	attempt := uint32(attemptValue)
	expected := webhookMAC(secret, timestamp, deliveryID, attempt, body)
	if subtle.ConstantTimeCompare(provided, expected) != 1 {
		return WebhookVerification{}, fmt.Errorf("epoch: webhook signature is invalid")
	}
	return WebhookVerification{DeliveryID: deliveryID, Attempt: attempt, SignedAt: signedAt}, nil
}

func webhookMAC(secret []byte, timestamp int64, deliveryID string, attempt uint32, body []byte) []byte {
	bodyDigest := sha256.Sum256(body)
	canonical := fmt.Sprintf(
		"v1\n%d\n%s\n%d\n%x",
		timestamp,
		deliveryID,
		attempt,
		bodyDigest,
	)
	mac := hmac.New(sha256.New, secret)
	_, _ = mac.Write([]byte(canonical))
	return mac.Sum(nil)
}

File: sdk/python/src/epoch_sdk/webhook.py
"""Verification helpers for Epoch signed webhook requests."""

from __future__ import annotations

import hashlib
import hmac
import time
from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class WebhookVerification:
    """Authenticated identity to persist before applying webhook side effects."""

    delivery_id: str
    attempt: int
    signed_at_seconds: int


def verify_webhook_signature(
    secret: bytes,
    body: bytes,
    delivery_id: str,
    attempt_header: str,
    timestamp_header: str,
    signature_header: str,
    *,
    now_seconds: int | None = None,
    tolerance_seconds: int = 300,
) -> WebhookVerification:
    """Verify Epoch's exact-body v1 signature and bounded replay timestamp."""

    if not isinstance(secret, bytes) or not secret:
        raise ValueError("webhook secret must be non-empty bytes")
    if not isinstance(body, bytes):
        raise TypeError("webhook body must be bytes")
    if not delivery_id.strip():
        raise ValueError("webhook delivery ID is required")
    attempt = _canonical_non_negative_integer(attempt_header, "attempt")
    if attempt == 0 or attempt > 4_294_967_295:
        raise ValueError("webhook attempt must be between 1 and 4294967295")
    timestamp = _canonical_non_negative_integer(timestamp_header, "timestamp")
    if isinstance(tolerance_seconds, bool) or tolerance_seconds <= 0:
        raise ValueError("webhook timestamp tolerance must be positive")
    if now_seconds is None:
        now_seconds = int(time.time())
    if isinstance(now_seconds, bool) or not isinstance(now_seconds, int):
        raise TypeError("webhook current time must be integer seconds")
    if abs(now_seconds - timestamp) > tolerance_seconds:
        raise ValueError("webhook timestamp is outside the allowed tolerance")
    if (
        len(signature_header) != 67
        or not signature_header.startswith("v1=")
        or signature_header[3:] != signature_header[3:].lower()
    ):
        raise ValueError("webhook signature must use v1 lowercase hexadecimal")
    try:
        provided = bytes.fromhex(signature_header[3:])
    except ValueError as error:
        raise ValueError("webhook signature must use v1 lowercase hexadecimal") from error
    body_digest = hashlib.sha256(body).hexdigest()
    canonical = f"v1\n{timestamp}\n{delivery_id}\n{attempt}\n{body_digest}".encode()
    expected = hmac.new(secret, canonical, hashlib.sha256).digest()
    if not hmac.compare_digest(provided, expected):
        raise ValueError("webhook signature is invalid")
    return WebhookVerification(delivery_id, attempt, timestamp)


def _canonical_non_negative_integer(value: str, name: str) -> int:
    if not value or not value.isascii() or not value.isdecimal():
        raise ValueError(f"webhook {name} must be a non-negative integer")
    parsed = int(value)
    if str(parsed) != value:
        raise ValueError(f"webhook {name} must use canonical decimal encoding")
    return parsed
